### ISSUE TITLE
Feature request: Build args on docker hub

### ISSUE THREAD

**Issue by cusspvz** [docker/hub-feedback#508]
Since we can have ARGS on dockerfile, seems right to have them available on auto-builds as also (for having different tags based on the same Dockerfile)

**Comment by cheungpat** [comment]
Together with LABEL, build args is very useful to apply custom metadata to images. For example:

```
FROM debian:wheezy

ARG branch
LABEL com.example.branch=$branch
```

Setting `--build-arg branch={sourceref}` can add the `com.example.branch=master` label to the image. It would be great if Docker Hub supports more build variables to provide metadata such as build id, git commit hash, creation time etc.

**Comment by kaendfinger** [comment]
Yes, it would be really useful. +404003

**Comment by andyneff** [comment]
I've been thinking the same thing since I started using buildargs. It would be perfect if there was a way to set specific buildargs for specific tags on my dockerhub image, instead of having to create separate tags/branches in my git repo.

**Comment by Yajo** [comment]
This would be even better than https://github.com/docker/hub-feedback/issues/292

**Comment by cusspvz** [comment]
@Yajo note that this is requesting access to `--build-args` option and not entire `docker build` cli args.

**Comment by Yajo** [comment]
Yes, I just meant that #292 could cover the same use cases, but fixing this would be much better for most of those.

**Comment by schickling** [comment]
This would make multi-version maintenance of images a lot easier

**Comment by pchico83** [comment]
This is available in Docker Cloud, which uses the same backend than Hub for autobuilds. The way to do it is to define autobuild env variables, an write a `build` hook that does the `docker build` command with the required build arguments.
https://docs.docker.com/docker-cloud/feature-reference/automated-build/
https://docs.docker.com/docker-cloud/feature-reference/automated-testing/

**Comment by nalipaz** [comment]
Can someone explain the method described by @pchico83? Because the documentation does not mention anything about a writing a `build` hook, only `hook/pre_build` and `hook/post_build` as part of testing from what I can tell.

**Comment by pchico83** [comment]
@nalipaz although it is not documented, it is supported (I will open an issue to improve the documentation). For example, if you define an env var called `VAR` and you want to set it to a build argument called TEST, you can define a build hook adding a `hooks/build` file to the dockerfile folder, whose content could be:
```
#!/bin/sh
docker build --build-arg TEST=$VAR -t $IMAGE_NAME .
```
Note that the hook is executed using the dockerfile folder as working dir, and `$IMAGE_NAME` is a variable we inject that corresponds with the docker tag being pushed.

**Comment by nalipaz** [comment]
That is much clearer, thanks @pchico83, and yes it would be very helpful to many I am sure were it documented. Thanks.

**Comment by Gurubol** [comment]
@pchico83 , Can you please provide me an example of how this hooks is used? Do we use the same docker file commands?
I tried with an example of setting an environment variable, which I later want to refer to in the Dockerfile, but it throws error

For example, in hooks/post_checkout, I added

ENV UI_BUILD x-1.2.2

But this throws me the following log.
"Executing post_checkout hook... 
Unexpected error"
Can you please throw some lights on this? Any help is appreciated.

**Comment by nalipaz** [comment]
Yeah, a gist or example project would likely be helpful for this. I will see if I can find time to test and make one although I am not using this feature currently. I opted for using Travis CI to use a make file to pass in params to the build and then push the final image into docker hub.

**Comment by pchico83** [comment]
@Gurubol You are probably missing the `bash shebang` at the beginning of your hook.
This is an example project with hooks:
https://github.com/docker/dockercloud-events

**Comment by andyneff** [comment]
Thanks to @pchico83 guidance, I was able to look around and discovered a number of useful pieces of information

1. The docker image used in docker hub to build dockers is called [docker/highland_builder](https://hub.docker.com/r/docker/highland_builder/tags/2.19.1_rc1/)
2. The majority of the hidden hooks features are come from the [builder.py](https://github.com/andyneff/highland_builder/blob/master/builder.py)
3. Here is a simple hooks [example](https://github.com/andyneff/docker_hook_test/tree/2340d7f9bf8791fc1f873faf0206f8a58925c195)
4. There are a number of key environment arguments set upon launching a build script, all of which you can use in your hooks.

```
BUILD_PATH=/
MAX_LOG_SIZE=67108864
SOURCE_BRANCH=master
DOCKER_REPO=index.docker.io/andyneff/docker_hook_test
DOCKERCFG=REDACTED
LOGS_POST_SPEC=REDACTED
BUILD_CODE=b9manhaqhfyqtmdqupcjyuq
README_POST_SPEC=REDACTED
SOURCE_TYPE=git
DOCKER_HOST=unix:///var/run/docker.sock
DOCKERFILE_POST_SPEC=REDACTED
DOCKER_TAG=latest
PUSH=true
PYTHONUNBUFFERED=1
SOURCE_URL=https://github.com/andyneff/docker_hook_test.git
DOCKERFILE_PATH=
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
DOCKER_VERSION=1.11.2
COMPOSE_VERSION=1.7.1
```

Redacted fields contained what looked like keys I assume would be harmless and expired by now, but I'm not sure ;)

But the take away is also have access to `SOURCE_BRANCH` and `DOCKER_TAG`, in addition to `IMAGE_NAME`, `GIT_SHA1`, `GIT_MSG`, and `COMMIT_MSG`, which can all be useful in making custom build-args :)

Hope this helps with this pretty awesome undocumented feature!

**Comment by thibaultdelor** [comment]
Here is a test image I use to check which hook are triggered when and which variables are available :
https://github.com/thibaultdelor/testAutobuildHooks

Check the result of the build here :
https://hub.docker.com/r/thibaultdelor/testautobuildhooks/builds/

**Comment by andschwa** [comment]
I *really* need this. I have multiple Dockerfiles because I have multiple distros to support, so they're each in their own folder, meaning they don't have the whole repo in their context, so the repo is cloned during `docker build` and I just need to be able to pass `﻿⁠⁠⁠⁠--build-arg branch=$(git rev-parse HEAD)` in the automated build so that I can checkout the right commit. Argh.

**Comment by andyneff** [comment]
@andschwa I prefer to have multiple named Dockerfiles in the root of the repo and then just run `docker build -f myrepo.Dockerfile` and not need to copy the repo. Either way, you can achieve this using my post above. See [here](https://github.com/git-lfs/build-dockers) for a full example when we are doing almost exactly what you are referring to. Just customize the `hooks/build` to your hearts content. Hope this helps

**Comment by zixia** [comment]
need this feature for the automatically build, will be really nice.

**Comment by opsxcq** [comment]
+1

**Comment by wangkuiyi** [comment]
+1

**Comment by tommed** [comment]
I'm in the same boat as @andschwa, we have build profiles and need to steer this via the code commit. Ideally, we'd love to get all the meta data surrounding the github/bitbucket webhook posted to us in environment variables or (better yet) buildargs! That way we can decide how to customise the build process per automated build.

**Comment by andyneff** [comment]
Something I forgot to mention earlier, but because of [this](https://github.com/andyneff/highland_builder/blob/3a588d28c5b964caa18827713d73ac08bf4d005e/builder.py#L284) line, a Dockerfile MUST exist at the "Dockerfile Location" column in the "Build settings". 

So if you are using a filename other than `./Dockerfile`, even though your hook will know what to do, the build script will erroneously fail. You must either:

- Have "Dockerfile Location" point to a directory in the git repo containing a file named `Dockerfile`
- Have "Dockerfile Location" point to a file in the git repo that will exist
- Create a dummy `Dockerfile` (this is the least ideal solution for me)

Also, don't forget to [give your `hooks/*` files 755 permissions](http://blog.lesc.se/2011/11/how-to-change-file-premissions-in-git.html). 

**Update**

Your `hooks` directory must be in the "Dockerfile Location" directory

**Comment by gbolo** [comment]
Really need this feature to avoid having a ton a dockerfiles to produce similar images. +1

**Comment by ju5t** [comment]
Nearly 2 years since this was originally requested. It would make life so much easier if this was possible. A definite +1 for me.

**Comment by Dean-Christian-Armada** [comment]
It would be clearly convenient if no hooks are needed and just clearly set the environment variable. With the DockerCloud UI, it really is hard to tell if hooks are needed and first impression would be just setting in the ENVIRONMENT VARIABLES  on the CONFIGURE AUTOMATED BUILDS page will be enough to act as build arguments

**Comment by felixcheung** [comment]
ARG with default value seems ok then? just no way to change them as I understand from this?

**Comment by geekgonecrazy** [comment]
I think the idea is you'd be able to set via webhook to override those set in the Dockerfile

**Comment by col-panic** [comment]
Is there any chance that this request will be realized? I am still waiting to access `SOURCE_BRANCH` as variable without explicitely setting it!

**Comment by alexeyzimarev** [comment]
@col-panic `SOURCE_BRANCH` is there https://docs.docker.com/docker-cloud/builds/advanced/

**Comment by mitar** [comment]
@alexeyzimarev Are those available in Docker Hub as well?

**Comment by norio-nomura** [comment]
I'm using `SOURCE_BRANCH` variable in `hooks/post_checkout` script on Docker Hub's automated build as https://hub.docker.com/r/norionomura/swiftlint/.

**Comment by mitar** [comment]
So those are available only inside hooks script, or also inside the building container?

**Comment by norio-nomura** [comment]
Yup, I'm rewriting them in [hooks/post_checkout](https://github.com/norio-nomura/docker-swiftlint/blob/master/swift-4/hooks/post_checkout) before the build starts.
The actual [build log](https://hub.docker.com/r/norionomura/swiftlint/builds/bdljyveydvt27awfrjnrmzr/) for `0.25.0` tag:
```log
…
Removing intermediate container 888ed5b40b95

Step 3/6 : ENV SWIFTLINT_REVISION "0.25.0"

 ---> Running in 07721eb2bae9
…
```

**Comment by norio-nomura** [comment]
[The documentation](https://docs.docker.com/docker-cloud/builds/advanced/#build-hook-examples) said that `hooks/build` can be overridden as:
```
docker build --build-arg SOURCE_BRANCH=$SOURCE_BRANCH -f $DOCKERFILE_PATH -t $IMAGE_NAME .
```
I don't know whether overriding `hooks/build` works on Docker Hub's automated build or not.

**Comment by col-panic** [comment]
> I don't know whether overriding hooks/build works on Docker Hub's automated build or not.

That would be really interesting to know! Thanks for your insight @norio-nomura

**Comment by nouchka** [comment]
It works, it's a workaround for this issue [https://github.com/nouchka/docker-symfony/blob/master/hooks/build](url)

**Comment by smartystreets-frontend** [comment]
Any movement on this? I have private NPM packages. I'd love to be able to inject auth into my automated builds so the images can build on push.

**Comment by cbuchert** [comment]
Any movement on this? My containers use private NPM packages and I'd love to be able to inject auth for NPM into my automated builds on push.

**Comment by browserslab-md5** [comment]
Either I am missing something from the docker hub documentation or this is no way to set my token for my private NPM packages. If anyone else has a solution for it, please post it. I am still wondering how no one has run into this issue.

**Comment by wshamim** [comment]
Just sharing my experience about how I get Build Arguments (Authentication, Branch) working for **Cloud Build** using private github repo.

1. Use **cloud.docker.com** instead of **hub.docker.com**
2. Go to <repository> => "Builds" => "Configure Automated Builds"
3. As my `Dockerfile` is inside my `<project root>/docker` directory, I updated `Dockerfile Location` to `/docker/Dockerfile`
4. Add 2 Environment Variables as `AUTH` and `BRANCH`
5. Create a `hooks` directory where my docker file is and create a build hook file like `<project root>/docker/hooks/build`
6. Update that `hooks/build` file with this code. 
[Build hook for docker cloud](https://gist.github.com/wshamim/b3f31a71e4bf9df0ccb3f0297fad9dcd)

That's it. So much trouble for just passing those Env Variables as Arguments. :)

Note: You build might fail on Docker Hub but it will compile and pass the arguments on Docker Cloud. But you will see the Tags getting updated on Docker Hub too.

**Comment by jcobian** [comment]
> Either I am missing something from the docker hub documentation or this is no way to set my token for my private NPM packages. If anyone else has a solution for it, please post it. I am still wondering how no one has run into this issue.

@ghost, I think I might have a workaround -- I was able to set my private NPM packages as git submodules in the repos where I need them. So I first re-installed the package from the local file system (from the submodule), so now my package.json file is using a relative path (note, be on npm >= 5 [to get relative paths instead of absolute paths](https://github.com/npm/npm/issues/15778#issuecomment-309033257)).

Then, the DockerHub auto build process can also install my private npm package from the local filesystem. You can ensure that the Docker Hub auto build process has access to submodules by following [this documentation](https://docs.docker.com/docker-hub/github/#github-private-submodules)

**Comment by cusspvz** [comment]
@Yajo this is for docker hub builds, not local/self-hosted docker builders.

**Comment by Yajo** [comment]
Yes, that's a Docker Hub automated build: https://hub.docker.com/r/tecnativa/duplicity/

**Comment by LinusU** [comment]
I have two flavors of my build, `stable` and `nightly`, and only need to switch a single build arg in order to produce both. It seems like the workaround only works for passing in some build args, but not actually running more builds than one, or am I missing something? 🤔 

Basically, I would like to run `docker build --build-arg FLAVOR=stable` *and* `docker build --build-arg FLAVOR=nightly` on each push to the repo...

**Comment by Chumper** [comment]
@LinusU I read in the documentation that only one run is allowed within 5 minutes.
Any other request will be ignored.

**Comment by opsxcq** [comment]
@LinusU your use case is that you want the build to produce more than one artifact (image). I totally agree with that, is fair common to have several images to be built in the same project, sometimes for other architectures, sometime several components (frontend/backend/etc...), and people don't want to split it into several small projects.

I agree with the importance of your request, some people may suggest that you use different git branches to archive that, but the flexibility to create several artifacts in one build is more interesting.

And happy birthday issue !!! 3 years old !!!

![img](http://thesoundofhope.org/wp-content/uploads/2013/05/3bday2.jpg)

**Comment by edify42** [comment]
> @nalipaz although it is not documented, it is supported (I will open an issue to improve the documentation). For example, if you define an env var called `VAR` and you want to set it to a build argument called TEST, you can define a build hook adding a `hooks/build` file to the dockerfile folder:
> 
> ```
> .
> ..
> hooks/build
> Dockerfile
> (other files)
> ```
> 
> whose content could be:
> 
> ```
> #!/bin/sh
> docker build --build-arg TEST=$VAR -t $IMAGE_NAME .
> ```
> 
> Note that the hook is executed using the dockerfile folder as working dir, and `$IMAGE_NAME` is a variable we inject that corresponds with the docker tag being pushed.

I think this is the fix and the issue should be closed?

**Comment by khalilgharbaoui** [comment]
@edify42 Well it could be closed, but the workaround does not actually supply the requested feature.

Right now it is possible to add `BUILD ENVIRONMENT VARIABLES`. But, in reality, those aren't really only for `build`... they are just `usual` ENV variables.

![image](https://user-images.githubusercontent.com/8024057/57961427-2f818b00-790f-11e9-8278-ad26b1fa5795.png)

This naming is very confusing IMO and it should be renamed to `ENVIRONMENT VARIABLES` instead.

And, in addition, we need a `BUILD ARGUMENTS` section where those can be filled in.

Because there is a clear difference between the two: https://vsupalov.com/docker-arg-vs-env/

The workaround with hooks is great and works fine, but it adds overhead.

**Comment by bitsofinfo** [comment]
2015... and this still isn't available?

**Comment by abdennour** [comment]
since 2015 till 2019/09/26 .. still not available ! :(

**Comment by mitar** [comment]
They are available, you just have to use build hooks.

**Comment by mitar** [comment]
https://docs.docker.com/docker-hub/builds/advanced/

**Comment by edify42** [comment]
> https://docs.docker.com/docker-hub/builds/advanced/

See my example...

**Comment by kingdonb** [comment]
Does this need to still be open? I see some thumbs down, but https://docs.docker.com/docker-hub/builds/advanced/#override-the-build-phase-to-set-variables actually solves this on Docker Hub. I was able to build an image with `RAILS_MASTER_KEY` set at build time, so I could run

```
RAILS_ENV=production bundle exec rake assets:precompile
```

and it worked absolutely fine! No sign of my build arg in the image output, which is the point of build-arg as I understand it. So, how are these `ENV` variables and not `ARG` variables? There is no sign of them in my build context either, unless I invoke the `ARG` keyword (in which case they are silently made available and not included in the image output, from what I can tell.)

To put a finer point on it, https://stackoverflow.com/a/55839775/661659

If you follow this example exactly, but in "step3" except instead of using `ARG` followed by `ENV` you simply call `ARG` with the name of the build-arg that you added into the hooks/build file, (per the same example previously linked by @edify42), you get a build-time only variable, it seems, unless there is something that I missed.

~@khalilgharbaoui can you clue me in if there's something I didn't understand here, or can the maintainers feel free to close this? I think it's solved.~

agh, maybe not quite secret, I guess this is what you meant:

```
<missing>           7 minutes ago       |1 RAILS_MASTER_KEY=c06afc06a31b6d17bf62361f…   180MB
<missing>           13 minutes ago      /bin/bash -l -c #(nop)  ARG RAILS_MASTER_KEY    0B
```

Is there a way to do this as we intended in Docker at all, or is it a request against nature?

I personally really did think this was an impossible request until I found numerous posts describing ARG as a way to put secrets into the build without letting them leak out through the images. But here it is, (at least most of) the secret (edit: yes, you can recover the whole secret and any other ARGs using `docker history --no-trunc`, here I see them in my `docker history` for the image, which I just built on Docker hub and pulled from there onto my docker machine.

It looks like they are leaked out through docker history, but not into the container image runtime context. "So I've got that going for me at least, which is nice."

TBH this might still be marked as solved, I'm not sure how Docker images can maintain a consistent format if they allow secrets to be used at build time but not included in `docker history` output reports. This strategy prevents them from being available inside of the image at runtime, which if you want those secrets to remain private you probably TBH shouldn't publish images that are based on them to public registries anyway. Thoughts?

**Comment by JuliusSustarevas** [comment]
Just wanna check what the current situation is:

- You can have a hooks/build where you "overwrite" the build command and can set your worn --build-arg.

But this method also overwrites **the whole** build command. Which means it also overwrites the --tag,--file, and most importantly **--no-cache** 



- So if I pass build-arg by having a custom hooks/build. The button  "use cashe" on "configure autobuilds"  on docker hub, will do nothing?

**Comment by HarelM** [comment]
I have read every message in this thread and it's still not clear to me if this is a problem in my code or a bug in dockerhub.
My use case is as follows (a very straight one from my point of view):
I set up two automation build in docker hub:
![image](https://user-images.githubusercontent.com/3269297/119785440-dd997580-bed7-11eb-8ec0-3a2fbac91311.png)

I'm expecting to be able to use the DOCKER_TAG within a non custom build and it seems to be empty.
My Dockerfile code is something like:
```
ARG DOCKER_TAG=v9.9.0
...
RUN if [ "$DOCKER_TAG" = "latest" ] || [ "$DOCKER_TAG" = "" ] ; then \
    echo "Building latest: $DOCKER_TAG" && dotnet build; \
    else \
    echo "Building version: ${DOCKER_TAG}" && dotnet build -p:"Version=${DOCKER_TAG:1};AssemblyVersion=${DOCKER_TAG:1}"; \
    fi
```
It seems that in both builds the DOCKER_TAG is empty...
Am I doing something wrong or is this a bug?

**Comment by andyneff** [comment]
It's been a while since I looked at this, but I believe `DOCKER_TAG` is an environment variable, not a "docker build arg".

By environment variable, I mean it is an environment variable set in the shell that calls `docker build`. As you may know, environment variables are not just "passed along" to `docker build`, you have to manually specify what gets passed along by using "docker build args" that are usually set to hard coded values, or variables/environment variables.

In order to manually pass along an environment variable like `DOCKER_TAG` you need to write your own `hooks/build` script, like in the example [here](https://github.com/andyneff/docker_hook_test/blob/2340d7f9bf8791fc1f873faf0206f8a58925c195/hooks/build)

So instead of saying:

```
docker build -t $IMAGE_NAME .
```

You would want something like

```
docker build --build-arg=DOCKER_TAG="${DOCKER_TAG}" -t $IMAGE_NAME .
```

Only then will the environment variable `DOCKER_TAG` be passed along as the build arg variable `DOCKER_TAG`

**Note**: `--build-arg=FOO="${BAR}"` would pass the environment variable `BAR` into the docker build arg variable named `FOO`
**Note**: `--build-arg=FOO="stuff"` would set the docker build arg variable named `FOO` to the value "stuff"

This feature request is about adding the ability to specify a value from the webpage dockerhub and have it passed along _somehow_ as a build arg. So the desire of this feature request is to set "stuff" from the webpage and have it set to the docker build arg "FOO", without needing to write your own `hooks/build` script, which is little different from what you are after.

For now @HarelM , you should be able to accomplish what you want specifically by using a `hooks/build` script example I gave you.

**Comment by HarelM** [comment]
@andyneff Thanks for the info!
Well... in that case I rather just use my CI to build and publish the docker image instead of trying to understand another partial CI environment which needs all kind of hooks and what not.
I would expect these exiting environment variables to be passed to the docker image build flow and that not everyone would need to create a custom build hook for this feature, but I guess I'm in the minority here... :-)
I understand there's a workaround but seeing that a lot of people need to use this workaround, and it feels like it should be part of the build system, again, in my Opinion...

**Comment by gpaluk** [comment]
I'm laughing because DockerHub... You had one job! Really?.. Why are the key value pair fields sat on the DockerHub auto-build configuration page? This has been open since 2015 and people are confused between setting those key value pairs and the notion of adding build variables via `--build-args`. I'm lost for words here tbh.

**Comment by cusspvz** [comment]
This request is open for more than 10 years. To me, it is just one example of why Docker Hub has been losing its dominance in both image building and hosting. I don't use the Docker Hub for neither building or as a registry for a long time...
